---
title: "mtDNA-Server 2"
output:
  flexdashboard::flex_dashboard:
    orientation: rows
    navbar:
      - { title: "About", href: "https://mitoverse.i-med.ac.at/#!pages/contact", align: right }
params:
  haplogroups: ../tests/data/report/haplogroups.txt
  haplocheck: ../tests/data/report/haplocheck.txt
  variants:  ../tests/data/report/variants_ann.txt
  statistics:  ../tests/data/report/sample_statistics.txt
  mapping:  ../tests/data/report/sample_mappings.txt  
  excluded_samples:  ../tests/data/report/excluded_samples.txt
  pipeline_parameters:  ../tests/data/report/params.txt
---

```{r setup, include=FALSE}


MIN_COVERAGE_PERCENTAGE = 90;
MIN_MEAN_BASE_QUALITY = 15;
MIN_MEAN_DEPTH = 50

library(ggplot2)
library(DT)
library(tidyverse)
library(data.table)
library(plotly)
library(flexdashboard)
library(knitr)

knitr::opts_chunk$set(echo = FALSE)
```

```{r echo=FALSE, warning=FALSE, error=FALSE, include=FALSE}

link_to_sample <- function(sample) {
  paste0('<a target=_blank href=samples/', sample, '.html>', sample,'</a>' )
}

variants <- read.delim(params$variants)
contamination <- read.delim(params$haplocheck)
haplogroups <- read.delim(params$haplogroups)
statistics <- read.delim(params$statistics)
pipeline_params <- read.delim(params$pipeline_parameters)

excluded_samples <- try(read.delim(params$excluded_samples, header = FALSE))
if(inherits(excluded_samples, "try-error")) {
  excluded_samples = data.frame(V1=c())
}


mapping <- read.delim(params$mapping)
mapping <- mapping %>%
  rename(Sample_Label = Sample)
mapping <- mapping %>%
  mutate(
    qc = case_when(
        Filename %in% excluded_samples$V1 ~ "FAILED",
        TRUE ~ "PASSED"
      )
  )

statistics <- spread(statistics, key = Parameter, value = Value)
statistics$MeanCoverage <- as.numeric(statistics$MeanDepth)
statistics$MeanBaseQuality <- as.numeric(statistics$MeanBaseQuality)
statistics <- merge(statistics, mapping, by.x="Sample", by.y="Filename")


haplogroups <- merge(haplogroups, mapping, by.x="SampleID", by.y="Filename") %>%
  arrange(Sample_Label)

contamination <- merge(contamination, mapping, by.x="Sample", by.y="Filename") %>%
  arrange(Sample_Label)

contaminated_samples = contamination %>%
  rename("Contamination_Status" = "Contamination.Status") %>%
  filter(Contamination_Status == "YES");

variants_count <- data.table(t(table(variants$ID)))
variants_count <- merge(variants_count, mapping, by.x="V2", by.y="Filename") %>%
  arrange(Sample_Label)

variants <- merge(variants, mapping, by.x="ID", by.y="Filename") %>%
  arrange(Sample_Label)

variant_caller <- pipeline_params %>%
  filter(Parameter == "Variant Caller")

```


# Samples

Row
-----------------------------------------------------------------------

### Number of Samples

```{r}
valueBox(nrow(statistics), icon = "fa-users")
```

### Excluded Samples

```{r}
valueBox(nrow(excluded_samples), icon = "fa-filter")
```

### Variant Caller

```{r}
valueBox(paste(variant_caller$Value), icon = "fa-cog")
```


### Contaminated Samples

```{r}
valueBox(nrow(contaminated_samples), icon = "fa-flask")
```



Row
-------------------------------------
    
    
### Summary  {data-width=750}

```{r}

colors <- c("#28a745", "#dc3545")
names(colors) <- c("PASSED", "FAILED")

colors_contamination <- c("#dddddd","#28a745", "#dc3545")
names(colors_contamination) <- c("-", "NO", "YES")

statistics %>%
  mutate (
    Sample_Link = link_to_sample(Sample_Label),
    contamination = case_when(
      qc == "FAILED" ~ "-",
      Sample_Label %in% contaminated_samples$Sample_Label ~ "YES",
      TRUE ~ "NO"
    )
  ) %>%
  select("Sample_Link", "MeanDepth", "CoveragePercentage", "CoveredBases", "MeanBaseQuality", "MeanMapQuality", "qc", "contamination") %>%
  datatable(
    colnames=c("Sample", "Mean Coverage", "Covered Bases (%)", "Covered Bases", "Mean Base Quality", "Mean Mapping Quality", "QC", "Cont."),
    options = list(
      bPaginate = FALSE
    ),
    escape = FALSE,
    class = 'cell-border stripe'
  ) %>% 
  formatStyle(
    'qc',
    backgroundColor = styleEqual(c("PASSED", "FAILED"), colors),
    color = 'white'
  ) %>% 
  formatStyle(
    'contamination',
    backgroundColor = styleEqual(c("-", "NO", "YES"), colors_contamination),
    color = 'white'
  )
```


### Workflow parameters and Software versions {data-width=250}

```{r}
pipeline_params %>%
  kable()
```

Row
-------------------------------------
    

### Mean Coverage per sample

```{r}

ggplotly(
  ggplot(statistics) +
    geom_col(aes(x=Sample_Label, y=MeanCoverage, fill=qc)) +
    theme(axis.text.x=element_text(angle=+90)) + 
    geom_hline(yintercept=MIN_MEAN_DEPTH, color="black", linetype="dashed") +
    scale_fill_manual(values=colors) +
    xlab("Samples") +
    ylab("Mean Coverage") +
    labs(fill = "")
)
```

### Mean Base Quality per sample

```{r}
ggplotly(
  ggplot(statistics) +
    geom_col(aes(x=Sample_Label, y=MeanBaseQuality, fill=qc)) +
    theme(axis.text.x=element_text(angle=+90)) + 
    geom_hline(yintercept=MIN_MEAN_BASE_QUALITY, color="black", linetype="dashed") +
    scale_fill_manual(values=colors) +
    xlab("Samples") +
    ylab("Mean Base Quality") +
    labs(fill = "")
)
```



# Contamination

```{r}
contaminated_samples %>%
  mutate (Sample_Link = link_to_sample(Sample_Label)) %>%
  select("Sample_Link", "Contamination_Status", "Contamination.Level") %>%
  datatable(
    colnames=c("Sample", "Contamination","Level"),
    options = list(
      bPaginate = FALSE
    ),
    escape = FALSE,
    class = 'cell-border stripe'
  )
```

---

# Variants and Heteroplasmies

Row
-------------------------------------

### Summary

```{r}
variants_count %>%
  mutate (Sample_Link = link_to_sample(Sample_Label)) %>%
  select("Sample_Link", "N") %>%
  datatable(
    colnames=c("Sample", "Number of Variants"),
    options = list(
      bPaginate = FALSE
    ),
    escape = FALSE,
    class = 'cell-border stripe'
  )
```

Row
-------------------------------------

### Common Variants and Hetereoplasmies


```{r}
variants <- variants %>%
  mutate(Type = case_when(
    Type == "0/1" ~ "2",
    Type == "1/0" ~ "2",
    Type == "INDEL" ~ "3",
    TRUE ~ as.character(Type)
  ))

summary_data <- aggregate(  Sample_Label ~ Pos + Type , data = variants, FUN = length)
summary_data <- summary_data %>% filter(Sample_Label >= 1)
summary_data <- summary_data %>% mutate(
  label = case_when(
    Type == 1 ~ "Variant",
    Type == 2 ~ "Heteroplasmy",
    Type == 3 ~ "INDEL",
    TRUE ~ "Unknown"
  )
)
summary_data$AF<-summary_data$Sample_Label / max(length(unique(variants$Sample_Label)))
ggplotly(
  ggplot(summary_data, aes(x = Pos, y = AF, color=label)) +
    geom_point() +
    xlab("Position") +
    ylab("Allele Frequency") +
    labs(color = "Type")
)
```

---


# Haplogroups


Row
-------------------------------------

### Summary

```{r}
haplogroups %>%
  mutate (Sample_Link = link_to_sample(Sample_Label)) %>%
  select("Sample_Link", "Haplogroup", "Quality") %>%
  datatable(
    colnames=c("Sample", "Haplogroup", "Quality"),
    options = list(
      bPaginate = FALSE
    ),
    escape = FALSE,
    class = 'cell-border stripe'
  )
```

Row
-------------------------------------

### Quality per Sample

```{r}
ggplotly(
  ggplot(haplogroups) +
    geom_col(aes(y=Quality, x=Sample_Label), alpha=0.7) +
    geom_hline(yintercept=0.90, color="green", linetype="dashed") +
    geom_hline(yintercept=0.80, color="orange", linetype="dashed") +
    geom_hline(yintercept=0.70, color="red", linetype="dashed") +
    xlab("Samples") +
    ylab("Haplogroup Quality") +
    theme(axis.text.x=element_text(angle=+90))
)
```
