---
title: "mtDNA-Server 2"
output:
  flexdashboard::flex_dashboard:
    orientation: rows
params:
  haplogroups: ../tests/data/report2/haplogroups.txt
  haplocheck: ../tests/data/report2/haplocheck.txt
  variants:  ../tests/data/report2/variants_ann.txt
  statistics:  ../tests/data/report2/sample_statistics.txt
---

```{r setup, include=FALSE}
library(ggplot2)
library(DT)
library(tidyverse)
library(data.table)
library(plotly)
library(flexdashboard)

knitr::opts_chunk$set(echo = FALSE)
```

```{r echo=FALSE, results='asis'}

variants <- read.delim(params$variants)
contamination <- read.delim(params$haplocheck)
haplogroups <- read.delim(params$haplogroups)
statistics <- read.delim(params$statistics)


statTable<-spread(statistics, key = Parameter, value = Value)
statTable$MeanCoverage<-as.numeric(statTable$MeanDepth)
statTable$MeanBaseQuality<-as.numeric(statTable$MeanBaseQuality)

variantCount<-data.table(t(table(variants$ID)))
variantCount$SampleID<-sub('\\.bam$', '', variantCount$V2)
contamination$SampleID<-sub('\\.bam$', '', contamination$Sample) 
haplogroups$SampleNID<-sub('\\.bam$', '', haplogroups$SampleID) 

statsTableMerged<-merge(statTable, contamination, by.x="Sample", by.y="SampleID")
statsTableMergedVariant<-merge(statsTableMerged, variantCount, by.x="Sample", by.y="SampleID")
statsTableMergedVariantHaplo<-merge(statsTableMergedVariant, haplogroups, by.x="Sample", by.y="SampleNID")

statsTableMergedVariantHaplo <- statsTableMergedVariantHaplo %>% arrange(Sample)

contaminated_samples = statsTableMergedVariantHaplo %>% mutate("Contamination_Status" = "Contamination.Status") %>% filter(Contamination_Status == "YES");


```


# Samples

Row
-----------------------------------------------------------------------

### Number of Samples

```{r}
valueBox(nrow(haplogroups), icon = "fa-users")
```

### Excluded Samples

```{r}
valueBox(-1, icon = "fa-filter")
```

### Samples with Bad Quality

```{r}
valueBox(-1, icon = "fa-exclamation")
```


### Contaminated Samples

```{r}
valueBox(nrow(contaminated_samples), icon = "fa-flask")
```



Row
-------------------------------------
    
    
### Summary

```{r echo=FALSE, results='asis'}
statsTableMergedVariantHaplo %>%
  select("Sample", "MeanDepth", "CoveragePercentage", "CoveredBases", "MeanBaseQuality", "MeanMapQuality") %>%
  datatable(colnames=c("Sample", "Mean Coverage", "Covered Bases (%)", "Covered Bases", "Mean Base Quality", "Mean Mapping Quality"))
```

Row
-------------------------------------
    

### Mean Coverage per sample

```{r echo=FALSE, results='asis'}
ggplotly(
  ggplot(statTable) +
    geom_col(aes(x=Sample, y=MeanCoverage), alpha=0.7) +
    theme(axis.text.x=element_text(angle=+90)) + 
    xlab("Samples") +
    ylab("Mean Coverage")
)
```

### Mean Base Quality per sample

```{r echo=FALSE, results='asis'}
ggplotly(
  ggplot(statTable) +
    geom_col(aes(x=Sample, y=MeanBaseQuality), alpha=0.7) +
    theme(axis.text.x=element_text(angle=+90)) + 
    geom_hline(yintercept=20, color="red", linetype="dashed") +
    xlab("Samples") +
    ylab("Mean Base Quality")
)
```



# Contamination

```{r echo=FALSE, results='asis'}
contaminated_samples %>%
  select("Sample", "Contamination_Status", "Contamination.Level") %>%
  datatable(colnames=c("Sample", "Contamination","Level"))
```

---

# Variants and Heteroplasmies

Row
-------------------------------------

### Summary

```{r echo=FALSE, results='asis'}
statsTableMergedVariantHaplo %>%
  select("Sample", "N") %>%
  datatable(colnames=c("Sample", "Number of Variants"))
```

Row
-------------------------------------

### Common Variants and Hetereoplasmies


```{r echo=FALSE, results='asis'}
summary_data <- aggregate(  ID ~ Pos + Type , data = variants, FUN = length)
summary_data <- summary_data %>% filter(ID >= 2)
summary_data <- summary_data %>% mutate(
  label = case_when(
    Type == 1 ~ "Variant",
    Type == 2 ~ "Heteroplasmy",
    TRUE ~ "Unknown"
  )
)
summary_data$AF<-summary_data$ID / max(length(unique(variants$ID)))
ggplotly(
  ggplot(summary_data, aes(x = Pos, y = AF, color=label)) +
    geom_point() +
    xlab("Position") +
    ylab("Allele Frequency") +
    labs(color = "Type")
)
```

---


# Haplogroups


Row
-------------------------------------

### Summary

```{r echo=FALSE, results='asis'}
statsTableMergedVariantHaplo %>%
  select("Sample", "Haplogroup","Quality") %>%
datatable()
```

Row
-------------------------------------

### Quality per Sample

```{r echo=FALSE, results='asis'}
ggplotly(
  ggplot(haplogroups) +
    geom_col(aes(y=Quality, x=SampleNID), alpha=0.7) +
    geom_hline(yintercept=0.90, color="green", linetype="dashed") +
    geom_hline(yintercept=0.80, color="orange", linetype="dashed") +
    geom_hline(yintercept=0.70, color="red", linetype="dashed") +
    xlab("Samples") +
    ylab("Haplogroup Quality") +
    theme(axis.text.x=element_text(angle=+90))
)
```



---

# About

### Software versions

table

### Workflow parameters

table

### About this report

This report was produced using mtDNA-Server 2. link...